nevergrad/functions/ttp/core.py

```diff
@@ -0,0 +1,224 @@
+
+            if line.startswith("RENTING RATIO"):
+                line = line[line.index(":") + 1 :]
+                line = line.strip()
+                self.rentingRatio = float(line)
+
+            if line.startswith("EDGE_WEIGHT_TYPE"):
+                line = line[line.index(":") + 1 :]
+                line = line.strip()
+                self.edgeWeightType = line
+
+            if line.startswith("NODE_COORD_SECTION"):
+                self.nodes = []
+
+                for i in range(0, self.numberOfNodes):
+                    count += 1 + i - i
+                    line = lines[count]
+                    splittedLine = line.split()
+                    col = []
+
+                    for j in range(0, len(splittedLine)):
+                        temp = float(splittedLine[j])
+                        if j == 0:
+                            temp = temp - 1
+                        col.append(temp)
+                    self.nodes.append(col)
+
+            if line.startswith("ITEMS SECTION"):
+                self.items = []
+                for i in range(0, self.numberOfItems):
+                    count += 1
+                    line = lines[count]
+                    splittedLine = line.split()
+                    col = []
+
+                    for j in range(0, len(splittedLine)):
+                        temp = int(splittedLine[j])
+                        # adjust city number by 1
+                        if j == 0:
+                            temp = temp - 1
+                        if j == 3:
+                            temp = temp - 1
+                        col.append(temp)
+                    self.items.append(col)
+
+            count += 1
+        reader.close()
+
+        init1 = np.zeros(self.numberOfNodes - 1)
+        p1 = p.Array(init=init1)
+        p1.set_bounds(0, 4)
+        p1.set_name("tspTour")
+
+        init2 = 0 * np.ones(self.numberOfItems)
+        p2 = p.Array(init=init2)
+        p2.set_integer_casting()
+        p2.set_bounds(0, 1)
+        p2.set_name("packingPlan")
+
+        instru = ng.p.Instrumentation(tspTour=p1, packingPlan=p2).set_name("")
+
+        super().__init__(self.evaluate, instru)
+
+    def evaluate(self, tspTour: np.ndarray, packingPlan: np.ndarray) -> float:
+
+        np.reshape(tspTour, (1, -1))
+        tspTour = np.argsort(tspTour)
+
+        tspTour = tspTour + 1
+        tspTL = tspTour.tolist()
+        tspTL.insert(0, 0)
+        tspTL.append(0)
+
+        packingPlan = packingPlan.tolist()
+
+        solution = TTPSolution(tspTL, packingPlan)
+
+        tour = solution.tspTour
+        z = solution.packingPlan
+        weightofKnapsack = self.capacityOfKnapsack
+        rentRate = self.rentingRatio
+        vmin = self.minSpeed
+        vmax = self.maxSpeed
+        solution.ftraw = 0
+
+        if tour[0] != tour[-1]:
+            print("ERROR: The last city must be the same as the first city")
+            solution.reset()
+            return float(-1)

```

Comment:
Is that supposed to be the worst loss? just to make sure we are aligned, nevergrad only minimizes, so a bad loss would be `math.inf` (or just a very large value)